spec/unit/mutant/isolation/fork_spec.rb

```diff
@@ -0,0 +1,264 @@
+        devnull: devnull,
+        io:      io,
+        marshal: marshal,
+        process: process,
+        stderr:  stderr,
+        stdout:  stdout
+      )
+    end
+
+    subject { object.call(&isolated_block) }
+
+    class MessageExpectation
+      include Anima.new(:receiver, :selector, :arguments, :return, :yields, :exception)
+
+      def self.new(receiver:, selector:, arguments: [], return: nil, yields: nil, exception: nil)
+        super
+      end
+
+      def call(observation)
+        Verifier.new(self, observation).call
+      end
+
+      class Verifier
+        include Concord.new(:expectation, :observation)
+
+        VERIFIED_ATTRIBUTES = IceNine.deep_freeze(%i[receiver selector arguments])
+
+        def call
+          VERIFIED_ATTRIBUTES.each(&method(:assert_expected_attribute))
+          trigger_exception
+          validate_block
+          expectation.return
+        end
+
+      private
+
+        def assert_expected_attribute(name)
+          error("#{name} mismatch") unless observation.public_send(name).eql?(expectation.public_send(name))
+        end
+
+        def error(message)
+          fail "#{message},\n observation:\n #{observation.inspect}\n expectation:\n #{expectation.inspect}"
+        end
+
+        def trigger_exception
+          exception = expectation.exception
+          fail exception if exception
+        end
+
+        def observation_block
+          observation.block
+        end
+
+        def expectation_yields
+          expectation.yields
+        end
+
+        def validate_block
+          return if expectation_yields.nil? && observation_block.nil?
+
+          if !expectation_yields ^ !observation_block
+            error('expectation of block yields does not match observation')
+          end
+
+          validate_block_arity
+        end
+
+        def validate_block_arity
+          expectation_arity, observation_arity = expectation_yields.length, observation_block.arity
+
+          unless expectation_arity.equal?(observation_arity)
+            error("block arity mismatch, expected #{expectation_arity} got #{observation_arity}")
+          end
+
+          observation_block.call(*expectation_yields)
+        end
+      end
+    end
+
+    class MessageObservation
+      include Anima.new(:receiver, :selector, :arguments, :block)
+    end
+
+    class ExpectationVerifier
+      include Concord.new(:expectations)
+
+      def call(observation)
+        expectation = expectations.shift or fail "No expected message but observed #{observation}"
+        expectation.call(observation)
+      end
+
+      def assert_done
+        expectations.empty? or fail "unconsumed expectations:\n#{expectations.map(&:inspect).join}"
+      end
+
+      # rubocop:disable MethodLength
+      def self.verify(rspec_context, expectations)
+        verifier = new(expectations)
+
+        hooks = expectations
+          .group_by { |expectation| [expectation.receiver, expectation.selector] }
```

Comment:
> I think #map, #uniq (or #to_set) might actually end up being more efficient.

I primarily intend to not produce garbadge to throw it away. But as I throw the values of the hash away I should not think about this.

Too bad there is no `uniq_with` or something.